lib/src/rules/use_function_declaration_instead_variable.dart

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.use_function_declaration_instead_variable;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:linter/src/analyzer.dart';
+
+const _desc = r'Use a function declaration to bind a function to a name.';
+
+const _details = r'''
+
+**DO** use a function declaration to bind a function to a name.
+
+**BAD:**
+```
+void main() {
+  var localFunction = () {
+    ...
+  };
+}
+```
+
+**GOOD:**
+```
+void main() {
+  localFunction() {
+    ...
+  }
+}
+```
+
+''';
+
+class UseFunctionDeclarationInsteadVariable extends LintRule {
+  _Visitor _visitor;
+  UseFunctionDeclarationInsteadVariable()
+      : super(
+            name: 'use_function_declaration_instead_variable',
+            description: _desc,
+            details: _details,
+            group: Group.style) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    if (node.rightHandSide.bestType is FunctionType) {
```

Comment:
You don't actually need to do either, but...

In analyzer, just like in Dart, the types are divided into `InterfaceType`s and `FunctionType`s. If a type is a `FunctionType`, then it's a function, but it's also a function if it's an `InterfaceType` that implements the interface `Function` or defined a method named `call`. The first code snippet does part of that work for you. The second, however, tests to see whether the type is exactly the `InterfaceType` representing the class `Function` (and will return false for subtypes of `Function`).